### PR TITLE
 fix: fix neumann bcs using constraints with convection terms

### PR DESCRIPTION
**Pull request by wd15**
This cleans up #714. Created a new PR as couldn't push to @atismer master branch.

Fix Neumann type boundary conditions using constraints for convection
terms. Previously the face value was not calculated using the constrained gradient
provided by the constraint requiring iterations to approach the
solution. The face value is now calculated using the constrained gradient, but
this is only first order accurate.

### CODE REVIEW

**Comment by wd15** [comment]
Looks like the tests are good apart from appveyor, which I having some sort of SSH issue

**Comment by wd15** [comment]
I'm keeping @atismer commits to give credit for the contribution although that code is completely overwritten.

**Comment by guyer** [comment]
AppVeyor gets wedged all the time. The only option is just rerun the failing jobs (I don't know if anybody but me has permission to do that, though).

File: fipy/terms/abstractConvectionTerm.py [hunk 1]
```diff
@@ -165,26 +165,84 @@ def _buildMatrix(self, var, SparseMatrix, boundaryConditions=(), dt=None, transi
...
             weight = self._getWeight(var, transientGeomCoeff, diffusionGeomCoeff)
 
             if 'implicit' in weight:
                 alpha = weight['implicit']['cell 1 diag']
             else:
                 alpha = 0.0
 
-            exteriorCoeff =  self.coeff * mesh.exteriorFaces
+            alpha_constraint = numerix.where(var.arithmeticFaceValue.constraintMask, 0.0, alpha)
+            alpha_constraint = numerix.where(var.faceGrad.constraintMask, 1.0, alpha)
```
Comment (guyer) [inline fipy/terms/abstractConvectionTerm.py]: Line 176 clobbers line 175

File: fipy/terms/abstractConvectionTerm.py [hunk 2]
```diff
@@ -165,26 +165,84 @@ def _buildMatrix(self, var, SparseMatrix, boundaryConditions=(), dt=None, transi
...
+        boundary condition with convection using constraints. This
+        tests that the Neumann boundary condition converges without
+        multiple iterations of the solve step. The Dirichlet boundary
+        condition is second order accurate. The Neumann is only first
+        order accurate.
+
+        >>> import fipy as fp
+        >>> import numpy as np
+
+        >>> def solve(nx, constrain_right):
```
Comment (guyer) [inline fipy/terms/abstractConvectionTerm.py]: Please stop writing Haskell in Python

Comment (wd15) [reply fipy/terms/abstractConvectionTerm.py]: Fixed now

File: fipy/terms/abstractConvectionTerm.py (hunk 1 above)
Comment (wd15) [reply fipy/terms/abstractConvectionTerm.py]: Removed it

File: fipy/terms/abstractConvectionTerm.py [hunk 3]
```diff
@@ -199,47 +198,70 @@ def divergence(face_value):
...
+        >>> var.constrain(1.0, mesh.facesLeft)
+        >>> var.constrain(0.0, mesh.facesRight)
+        >>> eqn = fp.CentralDifferenceConvectionTerm((1.0,)) - fp.DiffusionTerm()
+        >>> _ = eqn.sweep(var)
+        >>> expected =  (np.exp(mesh.x) - np.exp(1.)) / (1 - np.exp(1.))
+        >>> error1 = float(np.sqrt(((var.value - expected)**2 * mesh.dx).sum()))
+
+        >>> assert np.allclose(np.log(error1 / error0 )/ np.log(nx0 / nx1), 2.0, atol=0.02)
+
+        Test Nuemann boundary condition with a fixed value on the left
```
Comment (tkphd) [inline fipy/terms/abstractConvectionTerm.py]: Neumann

Comment (wd15) [reply fipy/terms/abstractConvectionTerm.py]: Too slow
